Realm/RLMSyncSession.mm

```diff
@@ -16,212 +16,107 @@
+#import "RLMSyncConfiguration_Private.hpp"
 #import "RLMSyncUser_Private.hpp"
-#import "RLMSyncUtil.h"
-#import "RLMTokenModels.h"
-#import "RLMUtil.hpp"
+#import "sync/sync_session.hpp"
 
-#import "sync_manager.hpp"
+using namespace realm;
 
-@implementation RLMSessionBindingPackage
-
-- (instancetype)initWithFileURL:(NSURL *)fileURL
-                     syncConfig:(RLMSyncConfiguration *)syncConfig
-                     standalone:(BOOL)isStandalone
-                          block:(RLMSyncBasicErrorReportingBlock)block {
-    if (self = [super init]) {
-        self.fileURL = fileURL;
-        self.syncConfig = syncConfig;
-        self.isStandalone = isStandalone;
-        self.block = block;
-        return self;
-    }
-    return nil;
+@interface RLMSyncSession () {
+    std::weak_ptr<SyncSession> _session;
 }
 
 @end
 
-@interface RLMSyncSession ()
-
-@property (nonatomic, readwrite) RLMSyncSessionState state;
-@property (nonatomic, readwrite) RLMSyncUser *parentUser;
-@property (nonatomic, readwrite) NSURL *realmURL;
-
-@property (nullable, nonatomic) RLMSyncSessionHandle *sessionHandle;
-
-@end
-
 @implementation RLMSyncSession
 
-- (instancetype)initWithFileURL:(NSURL *)fileURL realmURL:(NSURL *)realmURL {
+- (instancetype)initWithSyncSession:(std::shared_ptr<SyncSession>)session {
     if (self = [super init]) {
-        self.fileURL = fileURL;
-        self.realmURL = realmURL;
-        self.resolvedPath = nil;
-        self.deferredBindingPackage = nil;
-        _state = RLMSyncSessionStateUnbound;
+        _session = session;
         return self;
     }
     return nil;
 }
 
-- (nullable RLMSyncConfiguration *)configuration {
-    RLMSyncUser *user = self.parentUser;
-    if (user && self.state != RLMSyncSessionStateInvalid) {
-        return [[RLMSyncConfiguration alloc] initWithUser:user realmURL:self.realmURL];
+- (BOOL)isEqual:(id)object {
+    if (![object isKindOfClass:[RLMSyncSession class]]) {
+        return NO;
     }
-    return nil;
-}
-
-- (void)_logOut {
-    [self.sessionHandle logOut];
-    [self.refreshTimer invalidate];
-    _state = RLMSyncSessionStateLoggedOut;
-}
-
-- (void)_invalidate {
-    [self.refreshTimer invalidate];
-    _state = RLMSyncSessionStateInvalid;
-    self.sessionHandle = nil;
-    [self.parentUser _deregisterSessionWithRealmURL:self.realmURL];
-    self.parentUser = nil;
-}
-
-#pragma mark - per-Realm access token API
-
-- (void)configureWithAccessToken:(RLMServerToken)token
-                          expiry:(NSTimeInterval)expiry
-                            user:(RLMSyncUser *)user
-                          handle:(RLMSyncSessionHandle *)handle {
-    self.parentUser = user;
-    self.accessToken = token;
-    self.accessTokenExpiry = expiry;
-    self.sessionHandle = handle;
-    [self _scheduleRefreshTimer];
+    auto this_ptr = _session.lock();
+    auto that_ptr = ((RLMSyncSession *)object)->_session.lock();
+    return this_ptr && that_ptr && (this_ptr == that_ptr);
```

Comment:
This logic is a little odd. If you have two `RLMSyncSession`s that wrap the same session they will compare equal, but if the underlying session goes away the `RLMSyncSession`s will no longer compare equal?